### PR TITLE
Kubernetes/Openshift Service Discovery feature

### PR DESCRIPTION
**Pull request by guicassolato**
**What this PR does / why we need it**:

It implements a Service Discovery feature inspired by https://github.com/kubernetes/kubernetes/blob/release-1.4/docs/proposals/service-discovery.md using the cluster API.

**Which issue(s) this PR fixes** 

Related to https://github.com/3scale/porta/issues/34 ([THREESCALE-1363](https://issues.jboss.org/browse/THREESCALE-1363))

**Special notes for your reviewer**:
A few TODOs:
- [x] Simplify the UI so it contains only 2 select boxes: namespace/project and service name
- [x] Postpone the creation of the bare service, delegating it to the background job, rather than creating first and importing cluster data later
- [x] Fix retrieval of port info – use a specific annotation instead of choosing from spec/ports
- [x] Resolve spec type from the Content-Type instead of using "api.service.kubernetes.io/description-language"
- [ ] Add more tests 😜

Left for a future PR
- Send a 3scale notification by the end of the asynchronous process
- Make Service Discovery namespace (and other stuff) configurable
- Add lookup attribute to table `services` (based on k8s object’s selfLink?)
- Update/resync button
- Replace configurable access tokens with RHSSO (OAuth)

Depends on https://github.com/3scale/porta/pull/3

Specially in order to support updates to the ActiveDocs too, it would be nice to have an association between `Service` (API) and `ApiDoc::Service` in 3scale. I suggest doing it in a separate PR. _Update:_ https://github.com/3scale/porta/pull/62

### CODE REVIEW

File: app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee [hunk 1]
```diff
@@ -0,0 +1,59 @@
+$ ->
+  form_source = $('form#new_service_source')[0]
```
Comment (hallelujah) [inline app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee]: I'd stay with jQuery

File: app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+  clear_namespaces = ()->
+    clear_select_options form_discover.service_namespace
+
+  clear_service_names = ()->
+    clear_select_options form_discover.service_name
+
+  show_service_form = (source, discover_func, scratch_func)->
+    if source == 'discover'
+      form_scratch.classList.add('is-hidden')
```
Comment (hallelujah) [inline app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee]: If using jQuery like https://github.com/3scale/porta/pull/35/files#r221654192
`form_scratch.addClass('is-hidden')`

File: app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      form_discover.classList.remove('is-hidden')
```
Comment (hallelujah) [inline app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee]: Same here you could use `removeClass`

File: app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee (hunk 1 above)
Comment (hallelujah) [reply app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee]: If you do not want jQuery then `getElementById`

File: app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      discover_func() unless typeof discover_func == 'undefined'
+    else
+      form_scratch.classList.remove('is-hidden')
+      form_discover.classList.add('is-hidden')
+      scratch_func() unless typeof scratch_func == 'undefined'
```
Comment (hallelujah) [inline app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee]: This is cofee you can write it like `scratch_func?()`

File: app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee [hunk 5]
```diff
@@ -0,0 +1,59 @@
+$ ->
+  form_source = $('form#new_service_source')[0]
+  form_scratch = $('form#new_service,form#create_service')[0]
+  form_discover = $('form#service_discovery')[0]
+
+  clear_select_options = (select)->
+    $(select).find('option').not(':first').remove()
+
+  clear_namespaces = ()->
+    clear_select_options form_discover.service_namespace
```
Comment (hallelujah) [inline app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee]: Hmm so you create an attribute service_namespace?
Why not using `data` attribute?

File: app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee (hunk 1 above)
Comment (didierofrivia) [reply app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee]: `getElementById` might be better and we'd get rid of jquery

Comment (didierofrivia) [reply app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee]: well, we are using jquery extensively anyways

File: app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee [hunk 6]
```diff
@@ -14,13 +14,13 @@ $ ->
 
   show_service_form = (source, discover_func, scratch_func)->
     if source == 'discover'
-      form_scratch.classList.add('is-hidden')
-      form_discover.classList.remove('is-hidden')
+      $(form_scratch).addClass 'is-hidden'
```
Comment (guicassolato) [inline app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee]: I see you decided to keep it jQuery-ish. Why? Can `classList.add|remove` fail in any supported browser?

File: app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee [hunk 7]
```diff
@@ -14,13 +14,13 @@ $ ->
...
+      $(form_discover).removeClass 'is-hidden'
       discover_func() unless typeof discover_func == 'undefined'
```
Comment (guicassolato) [inline app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee]: `discover_func ?()`

File: app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+  fetch_services = (namespace)->
+    $.getJSON "/p/admin/service_discovery/namespaces/#{namespace}/services.json", (data)->
+      services = data.services
+      services.forEach (service, index, array) ->
+        addOptionToSelect(form_discover.service_name, service.metadata.name)
+      if services.length > 0
+        form_discover.service_name.selectedIndex = 1
+
+  if form_source != null and typeof form_source != 'undefined'
```
Comment (guicassolato) [inline app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee]: You can probably do `if form_source?`, no?

File: app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    $(form_source.source).click change_service_source
+
+  if form_discover != null and typeof form_discover != 'undefined'
```
Comment (guicassolato) [inline app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee]: See https://github.com/3scale/porta/pull/35/files#r223023135

File: app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee [hunk 10]
```diff
@@ -0,0 +1,72 @@
+$ ->
+  form_source = document.getElementById 'new_service_source'
+  form_scratch = document.querySelectorAll('form#new_service,form#create_service')[0]
+  form_discover = document.getElementById 'service_discovery'
+
+  clear_select_options = (select)->
+    options = select.getElementsByTagName('option')
+    i = options.length - 1
+    while i > 0
```
Comment (hallelujah) [inline app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js.coffee]: `option.remove() for option in options`
?

**Pull request merged**